File: src/main/java/com/refinedmods/refinedstorage/apiimpl/autocrafting/CraftingManager.java
package com.refinedmods.refinedstorage.apiimpl.autocrafting;

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingManager;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPattern;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternContainer;
import com.refinedmods.refinedstorage.api.autocrafting.craftingmonitor.ICraftingMonitorListener;
import com.refinedmods.refinedstorage.api.autocrafting.task.*;
import com.refinedmods.refinedstorage.api.network.INetwork;
import com.refinedmods.refinedstorage.api.network.INetworkNodeGraphEntry;
import com.refinedmods.refinedstorage.api.util.IComparer;
import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.apiimpl.autocrafting.task.v6.calculator.CalculationResult;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.Tag;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.items.IItemHandlerModifiable;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.*;

public class CraftingManager implements ICraftingManager {
    private static final int THROTTLE_DELAY_MS = 3000;

    private static final Logger LOGGER = LogManager.getLogger(CraftingManager.class);

    private static final String NBT_TASKS = "Tasks";
    private static final String NBT_TASK_TYPE = "Type";
    private static final String NBT_TASK_DATA = "Task";

    private final INetwork network;

    private final Map<Component, List<IItemHandlerModifiable>> containerInventories = new LinkedHashMap<>();
    private final Map<ICraftingPattern, Set<ICraftingPatternContainer>> patternToContainer = new HashMap<>();

    private final List<ICraftingPattern> patterns = new ArrayList<>();
    private final Map<FluidStackKey, ICraftingPattern> fluidPatternsByOutput = new HashMap<>();
    private final Map<ItemStackKey, ICraftingPattern> itemPatternsByOutput = new HashMap<>();

    private final Map<UUID, ICraftingTask> tasks = new LinkedHashMap<>();
    private final List<ICraftingTask> tasksToAdd = new ArrayList<>();
    private final List<UUID> tasksToCancel = new ArrayList<>();
    private final Map<Object, Long> throttledRequesters = new HashMap<>();
    private final Set<ICraftingMonitorListener> listeners = new HashSet<>();
    private ListTag tasksToRead;

    private static class FluidStackKey {
        private final Fluid fluid;
        private final CompoundTag tag;

        public FluidStackKey(FluidStack fluidStack) {
            this.fluid = fluidStack.getFluid();
            this.tag = fluidStack.getTag();
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            FluidStackKey that = (FluidStackKey) o;
            return fluid.equals(that.fluid) && Objects.equals(tag, that.tag);
        }

        @Override
        public int hashCode() {
            return Objects.hash(fluid, tag);
        }
    }

    private static class ItemStackKey {
        private final Item item;
        private final CompoundTag tag;

        public ItemStackKey(ItemStack itemStack) {
            this.item = itemStack.getItem();
            this.tag = itemStack.getTag();
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ItemStackKey that = (ItemStackKey) o;
            return item.equals(that.item) && Objects.equals(tag, that.tag);
        }

        @Override
        public int hashCode() {
            return Objects.hash(item, tag);
        }
    }

    public CraftingManager(INetwork network) {
        this.network = network;
    }

    @Override
    public Collection<ICraftingTask> getTasks() {
        return tasks.values();
    }

    @Override
    @Nullable
    public ICraftingTask getTask(UUID id) {
        return tasks.get(id);
    }

    @Override
    public Map<Component, List<IItemHandlerModifiable>> getNamedContainers() {
        return containerInventories;
    }

    @Override
    public void start(@Nonnull ICraftingTask task) {
        task.start();
        tasksToAdd.add(task);

        network.markDirty();
    }

    @Override
    public void cancel(@Nullable UUID id) {
        if (id == null) {
            tasksToCancel.addAll(tasks.keySet());
        } else {
            tasksToCancel.add(id);
        }

        network.markDirty();
    }

    @Override
    public ICalculationResult create(ItemStack stack, int quantity) {
        ICraftingPattern pattern = getPattern(stack);
        if (pattern == null) {
            return new CalculationResult(CalculationResultType.NO_PATTERN);
        }

        ICraftingTaskFactory factory = API.instance().getCraftingTaskRegistry().get(pattern.getCraftingTaskFactoryId());
        if (factory == null) {
            return new CalculationResult(CalculationResultType.NO_PATTERN);
        }

        return factory.create(network, API.instance().createCraftingRequestInfo(stack, quantity), quantity, pattern);
    }

    @Override
    public ICalculationResult create(FluidStack stack, int quantity) {
        ICraftingPattern pattern = getPattern(stack);
        if (pattern == null) {
            return new CalculationResult(CalculationResultType.NO_PATTERN);
        }

        ICraftingTaskFactory factory = API.instance().getCraftingTaskRegistry().get(pattern.getCraftingTaskFactoryId());
        if (factory == null) {
            return new CalculationResult(CalculationResultType.NO_PATTERN);
        }

        return factory.create(network, API.instance().createCraftingRequestInfo(stack, quantity), quantity, pattern);
    }

    @Override
    public void update() {
        if (network.canRun()) {
            if (tasksToRead != null) {
                readTasks();
            }

            boolean changed = !tasksToCancel.isEmpty() || !tasksToAdd.isEmpty();

            processTasksToCancel();
            processTasksToAdd();

            boolean anyFinished = updateTasks();

            if (changed || anyFinished) {
                onTaskChanged();
            }

            if (!tasks.isEmpty()) {
                network.markDirty();
            }
        }
    }

    private void processTasksToCancel() {
        for (UUID idToCancel : tasksToCancel) {
            if (this.tasks.containsKey(idToCancel)) {
                this.tasks.get(idToCancel).onCancelled();
                this.tasks.remove(idToCancel);
            }
        }
        this.tasksToCancel.clear();
    }

    private void processTasksToAdd() {
        for (ICraftingTask task : this.tasksToAdd) {
            this.tasks.put(task.getId(), task);
        }
        this.tasksToAdd.clear();
    }

    private boolean updateTasks() {
        boolean anyFinished = false;

        Iterator<Map.Entry<UUID, ICraftingTask>> it = tasks.entrySet().iterator();
        while (it.hasNext()) {
            ICraftingTask task = it.next().getValue();

            if (task.update()) {
                anyFinished = true;

                it.remove();
            }
        }

        return anyFinished;
    }

    private void readTasks() {
        for (int i = 0; i < tasksToRead.size(); ++i) {
            CompoundTag taskTag = tasksToRead.getCompound(i);

            ResourceLocation taskType = new ResourceLocation(taskTag.getString(NBT_TASK_TYPE));
            CompoundTag taskData = taskTag.getCompound(NBT_TASK_DATA);

            ICraftingTaskFactory factory = API.instance().getCraftingTaskRegistry().get(taskType);
            if (factory != null) {
                try {
                    ICraftingTask task = factory.createFromNbt(network, taskData);

                    tasks.put(task.getId(), task);
                } catch (CraftingTaskReadException e) {
                    LOGGER.error("Could not deserialize crafting task", e);
                }
            }
        }

        this.tasksToRead = null;
    }

    @Override
    public void readFromNbt(CompoundTag tag) {
        this.tasksToRead = tag.getList(NBT_TASKS, Tag.TAG_COMPOUND);
    }

    @Override
    public CompoundTag writeToNbt(CompoundTag tag) {
        ListTag list = new ListTag();

        for (ICraftingTask task : tasks.values()) {
            CompoundTag taskTag = new CompoundTag();

            taskTag.putString(NBT_TASK_TYPE, task.getPattern().getCraftingTaskFactoryId().toString());
            taskTag.put(NBT_TASK_DATA, task.writeToNbt(new CompoundTag()));

            list.add(taskTag);
        }

        tag.put(NBT_TASKS, list);

        return tag;
    }

    @Override
    public void addListener(ICraftingMonitorListener listener) {
        listeners.add(listener);

        listener.onAttached();
    }

    @Override
    public void removeListener(ICraftingMonitorListener listener) {
        listeners.remove(listener);
    }

    @Override
    public void onTaskChanged() {
        listeners.forEach(ICraftingMonitorListener::onChanged);
    }

    @Override
    @Nullable
    public ICraftingTask request(Object source, ItemStack stack, int amount) {
        if (isThrottled(source)) {
            return null;
        }

        for (ICraftingTask task : getTasks()) {
            if (task.getRequested().getItem() != null && API.instance().getComparer().isEqualNoQuantity(task.getRequested().getItem(), stack)) {
                amount -= task.getQuantity();
            }
        }

        if (amount > 0) {
            ICalculationResult result = create(stack, amount);

            if (result.isOk()) {
                start(result.getTask());

                return result.getTask();
            } else {
                throttle(source);
            }
        }

        return null;
    }

    @Nullable
    @Override
    public ICraftingTask request(Object source, FluidStack stack, int amount) {
        if (isThrottled(source)) {
            return null;
        }

        for (ICraftingTask task : getTasks()) {
            if (task.getRequested().getFluid() != null && API.instance().getComparer().isEqual(task.getRequested().getFluid(), stack, IComparer.COMPARE_NBT)) {
                amount -= task.getQuantity();
            }
        }

        if (amount > 0) {
            ICalculationResult result = create(stack, amount);

            if (result.isOk()) {
                start(result.getTask());

                return result.getTask();
            } else {
                throttle(source);
            }
        }

        return null;
    }

    private void throttle(Object source) {
        if (source != null) {
            throttledRequesters.put(source, System.currentTimeMillis());
        }
    }

    private boolean isThrottled(Object source) {
        if (source == null) {
            return false;
        }

        Long throttledSince = throttledRequesters.get(source);
        if (throttledSince == null) {
            return false;
        }

        return System.currentTimeMillis() - throttledSince < THROTTLE_DELAY_MS;
    }

    @Override
    public int track(@Nonnull ItemStack stack, int size) {
        if (stack.isEmpty()) {
            return 0;
        }

        for (ICraftingTask task : tasks.values()) {
            size = task.onTrackedInsert(stack, size);

            if (size == 0) {
                return 0;
            }
        }

        return size;
    }

    @Override
    public int track(@Nonnull FluidStack stack, int size) {
        if (stack.isEmpty()) {
            return 0;
        }

        for (ICraftingTask task : tasks.values()) {
            size = task.onTrackedInsert(stack, size);

            if (size == 0) {
                return 0;
            }
        }

        return size;
    }

    @Override
    public List<ICraftingPattern> getPatterns() {
        return patterns;
    }

    @Override
    public void invalidate() {
        this.network.getItemStorageCache().getCraftablesList().clear();
        this.network.getFluidStorageCache().getCraftablesList().clear();

        this.patterns.clear();
        this.fluidPatternsByOutput.clear();
        this.itemPatternsByOutput.clear();
        this.containerInventories.clear();
        this.patternToContainer.clear();

        List<ICraftingPatternContainer> containers = getContainers();

        for (ICraftingPatternContainer container : containers) {
            for (ICraftingPattern pattern : container.getPatterns()) {
                this.patterns.add(pattern);

                for (ItemStack output : pattern.getOutputs()) {
                    network.getItemStorageCache().getCraftablesList().add(output);
                    this.itemPatternsByOutput.put(new ItemStackKey(output), pattern);
                }

                for (FluidStack output : pattern.getFluidOutputs()) {
                    network.getFluidStorageCache().getCraftablesList().add(output);
                    this.fluidPatternsByOutput.put(new FluidStackKey(output), pattern);
                }

                Set<ICraftingPatternContainer> containersForPattern = this.patternToContainer.computeIfAbsent(pattern, key -> new LinkedHashSet<>());
                containersForPattern.add(container);
            }

            IItemHandlerModifiable handler = container.getPatternInventory();
            if (handler != null) {
                this.containerInventories.computeIfAbsent(container.getName(), k -> new ArrayList<>()).add(handler);
            }
        }

        this.network.getItemStorageCache().reAttachListeners();
        this.network.getFluidStorageCache().reAttachListeners();
    }

    private List<ICraftingPatternContainer> getContainers() {
        List<ICraftingPatternContainer> containers = new ArrayList<>();

        for (INetworkNodeGraphEntry entry : network.getNodeGraph().all()) {
            if (entry.getNode() instanceof ICraftingPatternContainer && entry.getNode().isActive()) {
                containers.add((ICraftingPatternContainer) entry.getNode());
            }
        }

        containers.sort((a, b) -> b.getPosition().compareTo(a.getPosition()));

        return containers;
    }

    @Override
    public Set<ICraftingPatternContainer> getAllContainers(ICraftingPattern pattern) {
        return patternToContainer.getOrDefault(pattern, Collections.emptySet());
    }

    @Nullable
    @Override
    public ICraftingPattern getPattern(ItemStack pattern) {
        return itemPatternsByOutput.get(new ItemStackKey(pattern));
    }

    @Nullable
    @Override
    public ICraftingPattern getPattern(FluidStack pattern) {
        return fluidPatternsByOutput.get(new FluidStackKey(pattern));
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/autocrafting/CraftingManager.kt
package com.refinedmods.refinedstorage.apiimpl.autocrafting

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingManager
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPattern
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternContainer
import com.refinedmods.refinedstorage.api.autocrafting.craftingmonitor.ICraftingMonitorListener
import com.refinedmods.refinedstorage.api.autocrafting.task.CalculationResultType
import com.refinedmods.refinedstorage.api.autocrafting.task.CraftingTaskReadException
import com.refinedmods.refinedstorage.api.autocrafting.task.ICalculationResult
import com.refinedmods.refinedstorage.api.autocrafting.task.ICraftingTask
import com.refinedmods.refinedstorage.api.network.INetwork
import com.refinedmods.refinedstorage.api.util.IComparer
import com.refinedmods.refinedstorage.apiimpl.API.Companion.instance
import com.refinedmods.refinedstorage.apiimpl.autocrafting.task.v6.calculator.CalculationResult
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.util.Identifier
import net.minecraft.util.text.Text
import net.minecraftforge.common.util.Constants
import net.minecraftforge.fluids.FluidInstance
import net.minecraftforge.items.IItemHandlerModifiable
import org.apache.logging.log4j.LogManager
import java.util.*
import java.util.function.Consumer


class CraftingManager(private val network: INetwork) : ICraftingManager {
    private val containerInventories: MutableMap<Text?, MutableList<IItemHandlerModifiable?>?> = LinkedHashMap<Text?, MutableList<IItemHandlerModifiable?>?>()
    private val patternToContainer: MutableMap<ICraftingPattern?, MutableSet<ICraftingPatternContainer>> = HashMap()
    private val patterns: MutableList<ICraftingPattern?> = ArrayList()
    private val tasks: MutableMap<UUID?, ICraftingTask?> = LinkedHashMap()
    private val tasksToAdd: MutableList<ICraftingTask?> = ArrayList()
    private val tasksToCancel: MutableList<UUID> = ArrayList()
    private var tasksToRead: ListTag? = null
    private val throttledRequesters: MutableMap<Any, Long> = HashMap()
    private val listeners: MutableSet<ICraftingMonitorListener?> = HashSet()
    override fun getTasks(): Collection<ICraftingTask?>? {
        return tasks.values
    }

    @Nullable
    override fun getTask(id: UUID?): ICraftingTask? {
        return tasks[id]
    }

    override fun getNamedContainers(): Map<Text?, MutableList<IItemHandlerModifiable?>?>? {
        return containerInventories
    }

    override fun start(@Nonnull task: ICraftingTask?) {
        task!!.start()
        tasksToAdd.add(task)
        network.markDirty()
    }

    override fun cancel(@Nullable id: UUID?) {
        if (id == null) {
            tasksToCancel.addAll(tasks.keys)
        } else {
            tasksToCancel.add(id)
        }
        network.markDirty()
    }

    override fun create(stack: ItemStack?, quantity: Int): ICalculationResult? {
        val pattern: ICraftingPattern = getPattern(stack) ?: return CalculationResult(CalculationResultType.NO_PATTERN)
        val factory = instance().getCraftingTaskRegistry()!![pattern.getCraftingTaskFactoryId()]
                ?: return CalculationResult(CalculationResultType.NO_PATTERN)
        return factory.create(network, instance().createCraftingRequestInfo(stack), quantity, pattern)
    }

    override fun create(stack: FluidInstance?, quantity: Int): ICalculationResult? {
        val pattern: ICraftingPattern = getPattern(stack) ?: return CalculationResult(CalculationResultType.NO_PATTERN)
        val factory = instance().getCraftingTaskRegistry()!![pattern.getCraftingTaskFactoryId()]
                ?: return CalculationResult(CalculationResultType.NO_PATTERN)
        return factory.create(network, instance().createCraftingRequestInfo(stack), quantity, pattern)
    }

    override fun update() {
        if (network.canRun()) {
            if (tasksToRead != null) {
                for (i in tasksToRead!!.indices) {
                    val taskTag = tasksToRead!!.getCompound(i)
                    val taskType = Identifier(taskTag.getString(NBT_TASK_TYPE))
                    val taskData = taskTag.getCompound(NBT_TASK_DATA)
                    val factory = instance().getCraftingTaskRegistry()!![taskType]
                    if (factory != null) {
                        try {
                            val task = factory.createFromNbt(network, taskData)
                            tasks[task!!.getId()] = task
                        } catch (e: CraftingTaskReadException) {
                            LOGGER.error("Could not deserialize crafting task", e)
                        }
                    }
                }
                tasksToRead = null
            }
            val changed = !tasksToCancel.isEmpty() || !tasksToAdd.isEmpty()
            for (idToCancel in tasksToCancel) {
                if (tasks.containsKey(idToCancel)) {
                    tasks[idToCancel]!!.onCancelled()
                    tasks.remove(idToCancel)
                }
            }
            tasksToCancel.clear()
            for (task in tasksToAdd) {
                tasks[task!!.getId()] = task
            }
            tasksToAdd.clear()
            var anyFinished = false
            val it: MutableIterator<Map.Entry<UUID?, ICraftingTask?>> = tasks.entries.iterator()
            while (it.hasNext()) {
                val task = it.next().value
                if (task!!.update()) {
                    anyFinished = true
                    it.remove()
                }
            }
            if (changed || anyFinished) {
                onTaskChanged()
            }
            if (!tasks.isEmpty()) {
                network.markDirty()
            }
        }
    }

    override fun readFromNbt(tag: CompoundTag?) {
        tasksToRead = tag!!.getList(NBT_TASKS, Constants.NBT.TAG_COMPOUND)
    }

    override fun writeToNbt(tag: CompoundTag?): CompoundTag? {
        val list = ListTag()
        for (task in tasks.values) {
            val taskTag = CompoundTag()
            taskTag.putString(NBT_TASK_TYPE, task!!.getPattern()!!.getCraftingTaskFactoryId().toString())
            taskTag.put(NBT_TASK_DATA, task.writeToNbt(CompoundTag()))
            list.add(taskTag)
        }
        tag!!.put(NBT_TASKS, list)
        return tag
    }

    override fun addListener(listener: ICraftingMonitorListener?) {
        listeners.add(listener)
        listener!!.onAttached()
    }

    override fun removeListener(listener: ICraftingMonitorListener?) {
        listeners.remove(listener)
    }

    override fun onTaskChanged() {
        listeners.forEach(Consumer { obj: ICraftingMonitorListener? -> obj!!.onChanged() })
    }

    @Nullable
    override fun request(source: Any?, stack: ItemStack?, amount: Int): ICraftingTask? {
        var amount = amount
        if (isThrottled(source)) {
            return null
        }
        for (task in getTasks()!!) {
            if (task!!.getRequested()!!.item != null) {
                if (instance().getComparer()!!.isEqualNoQuantity(task.getRequested()!!.item!!, stack!!)) {
                    amount -= task.getQuantity()
                }
            }
        }
        if (amount > 0) {
            val result: ICalculationResult = create(stack, amount)
            if (result.isOk()) {
                start(result.getTask())
            } else {
                throttle(source)
            }
        }
        return null
    }

    @Nullable
    override fun request(source: Any?, stack: FluidInstance?, amount: Int): ICraftingTask? {
        var amount = amount
        if (isThrottled(source)) {
            return null
        }
        for (task in getTasks()!!) {
            if (task!!.getRequested()!!.fluid != null) {
                if (instance().getComparer()!!.isEqual(task.getRequested()!!.fluid!!, stack, IComparer.COMPARE_NBT)) {
                    amount -= task.getQuantity()
                }
            }
        }
        if (amount > 0) {
            val result: ICalculationResult = create(stack, amount)
            if (result.isOk()) {
                start(result.getTask())
            } else {
                throttle(source)
            }
        }
        return null
    }

    private fun throttle(source: Any?) {
        if (source != null) {
            throttledRequesters[source] = System.currentTimeMillis()
        }
    }

    private fun isThrottled(source: Any?): Boolean {
        if (source == null) {
            return false
        }
        val throttledSince = throttledRequesters[source] ?: return false
        return System.currentTimeMillis() - throttledSince < THROTTLE_DELAY_MS
    }

    override fun track(@Nonnull stack: ItemStack?, size: Int): Int {
        var size = size
        if (stack!!.isEmpty) {
            return 0
        }
        for (task in tasks.values) {
            size = task.onTrackedInsert(stack, size)
            if (size == 0) {
                return 0
            }
        }
        return size
    }

    override fun track(@Nonnull stack: FluidInstance?, size: Int): Int {
        var size = size
        if (stack.isEmpty()) {
            return 0
        }
        for (task in tasks.values) {
            size = task.onTrackedInsert(stack, size)
            if (size == 0) {
                return 0
            }
        }
        return size
    }

    override fun getPatterns(): List<ICraftingPattern?>? {
        return patterns
    }

    override fun invalidate() {
        network.itemStorageCache!!.getCraftablesList()!!.clear()
        network.fluidStorageCache!!.getCraftablesList()!!.clear()
        patterns.clear()
        containerInventories.clear()
        patternToContainer.clear()
        val containers: MutableList<ICraftingPatternContainer> = ArrayList()
        for (node in network.nodeGraph!!.all()!!) {
            if (node is ICraftingPatternContainer && node.isActive) {
                containers.add(node as ICraftingPatternContainer)
            }
        }
        containers.sort(java.util.Comparator { a: ICraftingPatternContainer, b: ICraftingPatternContainer -> b.getPosition()!!.compareTo(a.getPosition()) })
        for (container in containers) {
            for (pattern in container.getPatterns()!!) {
                patterns.add(pattern)
                for (output in pattern!!.getOutputs()) {
                    network.itemStorageCache!!.getCraftablesList()!!.add(output)
                }
                for (output in pattern.getFluidOutputs()) {
                    network.fluidStorageCache!!.getCraftablesList()!!.add(output)
                }
                var list = patternToContainer[pattern]
                if (list == null) {
                    list = LinkedHashSet()
                }
                list.add(container)
                patternToContainer[pattern] = list
            }
            val handler: IItemHandlerModifiable? = container.getPatternInventory()
            if (handler != null) {
                containerInventories.computeIfAbsent(container.getName()) { k: Text? -> ArrayList<IItemHandlerModifiable?>() }!!.add(handler)
            }
        }
        network.itemStorageCache!!.reAttachListeners()
        network.fluidStorageCache!!.reAttachListeners()
    }

    override fun getAllContainers(pattern: ICraftingPattern?): Set<ICraftingPatternContainer?>? {
        return patternToContainer.getOrDefault(pattern, emptySet())
    }

    @Nullable
    override fun getPattern(pattern: ItemStack?): ICraftingPattern? {
        for (patternInList in patterns) {
            for (output in patternInList!!.getOutputs()) {
                if (instance().getComparer()!!.isEqualNoQuantity(output!!, pattern!!)) {
                    return patternInList
                }
            }
        }
        return null
    }

    @Nullable
    override fun getPattern(pattern: FluidInstance?): ICraftingPattern? {
        for (patternInList in patterns) {
            for (output in patternInList!!.getFluidOutputs()) {
                if (instance().getComparer().isEqual(output, pattern, IComparer.COMPARE_NBT)) {
                    return patternInList
                }
            }
        }
        return null
    }

    companion object {
        private const val THROTTLE_DELAY_MS = 3000
        private val LOGGER = LogManager.getLogger(CraftingManager::class.java)
        private const val NBT_TASKS = "Tasks"
        private const val NBT_TASK_TYPE = "Type"
        private const val NBT_TASK_DATA = "Task"
    }
}